core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -526,12 +561,21 @@ public IndexMetaData apply(IndexMetaData part) {
         }
     }
 
+    public static boolean doesClusterUUIDExistOnVersion(final Settings settings) {
+        Version indexVersion = settings.getAsVersion(IndexMetaData.SETTING_VERSION_UPGRADED, Version.indexCreated(settings));
+        return indexVersion.onOrAfter(Version.V_5_0_0);
+    }
+
     @Override
     public IndexMetaData readFrom(StreamInput in) throws IOException {
         Builder builder = new Builder(in.readString());
         builder.version(in.readLong());
         builder.state(State.fromId(in.readByte()));
         builder.settings(readSettingsFromStream(in));
+        if (doesClusterUUIDExistOnVersion(builder.settings)) {
```

Comment:
we only use toXContent for file persistency. Over the network we can also read uuids..